[Property wrappers] Fix handling of @autoclosure in init(wrappedValue:)

roop opened the pull request:
If the ‘wrappedValue:’ parameter is an escaping autoclosure, and a
struct property is marked with that property wrapper, the memberwise
initializer of the struct is now synthesized with an escaping
autoclosure for that property.

Resolves [SR-10950](https://bugs.swift.org/browse/SR-10950).

Given a Lazy property wrapper:
```
@propertyWrapper
enum Lazy<Value> {
  …
  init(wrappedValue: @autoclosure @escaping () -> Value) {
    …
  }
  …
```

and a struct using the property wrapper
```
struct S {
  @Lazy var n: Int = computeInt()
}
```

the memberwise initializer for S is currently synthesized as:
```
init(n: Int = computeInt())
```

This is incorrect because if we create S like `var s1 = S()` or like `var s2 = S(n: computeInt())`, `computeInt()` will get evaluated at the time of creation of S, which
  - is not consistent with the `@autoclosure` specification
  - makes `@Lazy` not really initialize things lazily

To fix this, this PR makes the memberwise initializer get generated as:
```
init(n: @autoclosure @escaping () -> Int = computeInt())
```

### Implementation details:

1. Changed the type of the opaque placeholder (_lib/Sema/TypeCheckStorage.cpp_): PropertyWrapperBackingPropertyInfo.underlyingValue is an opaque expression that serves as a placeholder for what should be used when initializing a struct using a property wrapper. For the above scenario, the type of this expression is changed to **() -> Int** (rather than **Int** previously).
2. Changed the signature of the memberwise initializer (_lib/Sema/CodeSynthesis.cpp_): The argument type is changed to a closure and autoclosure-ness is set.
3. Handled default values provided with an “=“ in the wrapper (_lib/SIL/TypeLowering.cpp_, _lib/SILGen/SILGenConstructor.cpp_, _lib/SILGen/SILGenFunction.cpp_): The type is likewise changed to a closure.
4. Because of (1), the opaque underlyingValue is of a closure type. I should ideally be passing this as the ‘wrappedValue:’ parameter, but we can’t pass a closure to an autoclosure arg. So, instead, I pass an AutoClosureExpr of a CallExpr of a call to this closure. (_lib/Sema/TypeCheckPropertyWrapper.cpp_)
5. Because of (4), I now have an extra CallExpr in the expression tree which causes problems in SILGen. So, while traversing the expression tree, I skip the AutoClosureExpr and CallExpr. (_lib/SILGen/SILGenExpr.cpp_)

roop commented:
@DougGregor @jckarter: Please review

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -5599,6 +5599,53 @@ bool VarDecl::isPropertyMemberwiseInitializedWithWrappedType() const {
 
+static bool isInitFirstParamEscapingClosure(const ConstructorDecl *init) {
+  Type initTy = init->getInterfaceType();
+  if (!initTy || !initTy->is<AnyFunctionType>())
```
Better to assume a type has been set here. It would be weird if it wasn't because then the result of the caller would depend on type checking order.

roop replied:
Okay. Removed the check in 1bc9dbd.

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -5599,6 +5599,53 @@ bool VarDecl::isPropertyMemberwiseInitializedWithWrappedType() const {
+    return false;
+
+  Type resultTy = initTy->castTo<AnyFunctionType>()->getResult();
+  if (!resultTy || !resultTy->is<FunctionType>())
```
Similarly, you can assume that structural components of types are not null.

roop replied:
Right, okay.

slavapestov commented on test/SILGen/property_wrappers.swift:
```diff
@@ -228,8 +228,8 @@ struct UseLazy<T: DefaultInit> {
-  // CHECK-LABEL: sil hidden [ossa] @$s17property_wrappers7UseLazyV3foo3bar6wibbleACyxGSi_xSaySiGtcfC : $@convention(method) <T where T : DefaultInit> (Int, @in T, @owned Array<Int>, @thin UseLazy<T>.Type) -> @out UseLazy<T>
-  // CHECK: function_ref @$s17property_wrappers7UseLazyV4_foo33_{{.*}}AA0D0OySiGvpfiSiycfu_ : $@convention(thin) (@owned Int) -> Int
+  // CHECK-LABEL: sil hidden [ossa] @$s17property_wrappers7UseLazyV3foo3bar6wibbleACyxGSiyXA_xyXASaySiGyXAtcfC : $@convention(method) <T where T : DefaultInit> (@owned @callee_guaranteed () -> Int, @owned @callee_guaranteed () -> @out T, @owned @callee_guaranteed () -> @owned Array<Int>, @thin UseLazy<T>.Type) -> @out UseLazy<T> {
+  // CHECK: function_ref @$sSiIegd_SiIegr_TR : $@convention(thin) (@guaranteed @callee_guaranteed () -> Int) -> @out Int
```
This needs more tests to demonstrate the behavioral change

roop replied:
Right. Added some tests in 1a5f57a. Do they look okay?

roop replied:
Also, I think there should be tests to ensure that:

 1. Only the innermost wrapper is considered 
 2. Only escaping closures are considered 

I’ll add them tomorrow.

theblixguy commented on lib/AST/Decl.cpp:
```diff
@@ -5599,6 +5599,53 @@ bool VarDecl::isPropertyMemberwiseInitializedWithWrappedType() const {
+  auto firstParamName = firstParam->getName();
+
+  ASTContext &ctx = getASTContext();
+  if (firstParamName != ctx.Id_wrappedValue &&
```
Shouldn't this be `||` instead of `&&`?

roop replied:
Why do you say that? As it stands, it reads: If the firstParamName is neither "wrappedValue" nor "initialValue", return false, which looks correct to me.

theblixguy replied:
Oops, I misread the check, sorry. The check looks correct.

roop commented:
In 1384ef0, for testing composition with autoclosureness, I'm not using @Lazy because it crashes, and that's unrelated to autoclosure handling. Specifically, it crashes if there's a property wrapper with a mutating get (like Lazy) being composed with another outer wrapper. So I'm using a simpler wrapper with autoclosure instead of Lazy in the test.

Cc: @DougGregor @jckarter

roop commented:
There were conflicts with the latest master, so I forced pushed after rebasing to the latest master.

roop commented:
@DougGregor @jckarter: Can you please review this PR? I think having `@autoclosure` working correctly is important for the implementation to correctly match SE-0258.

DougGregor commented:
@swift-ci please smoke test

slavapestov commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -2357,9 +2357,37 @@ RValue RValueEmitter::visitCaptureListExpr(CaptureListExpr *E, SGFContext C) {
+
+        Type aceType = ace->getType();
+        Type oveType = ove->getType();
+        if (!aceType || !oveType)
```
Expression types are never null

slavapestov commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -2357,9 +2357,37 @@ RValue RValueEmitter::visitCaptureListExpr(CaptureListExpr *E, SGFContext C) {
 RValue RValueEmitter::visitAbstractClosureExpr(AbstractClosureExpr *e,
                                                SGFContext C) {
+  if (auto ove = opaqueValueExprToSubstituteForAutoClosure(e))
+    return visitOpaqueValueExpr(ove, C);
```
It's not clear if this is an optimization or has semantic effect. Can you add a clarifying comment?

slavapestov reviewed (commented):
Just a few minor nitpicks. I'll defer to @DougGregor about the overall approach.

DougGregor commented on lib/AST/Decl.cpp:
```diff
@@ -5625,6 +5625,53 @@ bool VarDecl::isPropertyMemberwiseInitializedWithWrappedType() const {
+  if (!initTy->is<AnyFunctionType>())
+    return false;
+
+  Type resultTy = initTy->castTo<AnyFunctionType>()->getResult();
```
Just a minor stylistic thing: rather than `->is<AnyFunctionType>()` followed by `->castTo<AnyFunctionType>()`, you can use `->getAs<AnyFunctionType>()` and then check if the resulting pointer is NULL.

DougGregor commented on lib/SILGen/SILGenConstructor.cpp:
```diff
@@ -914,6 +914,10 @@ static Type getInitializationTypeInContext(
         interfaceType = originalProperty->getValueInterfaceType();
+        if (originalProperty->
+            isInnermostPropertyWrapperInitUsesEscapingAutoClosure()) {
+          interfaceType = FunctionType::get({}, interfaceType);
```
This check for `isInnermostPropertyWrapperInitUsesEscapingAutoClosure` followed by `FunctionType::get()` occurs in a few places, and it would be really easy to forget to do this adjustment anywhere else in the code. Can you add a new API on `VarDecl` that gets the interface type for wrapped value initialization, which does the `getValueInterfaceType()` and (now) the `FunctionType::get()`?

DougGregor commented on lib/SILGen/SILGenFunction.cpp:
```diff
@@ -696,6 +696,11 @@ void SILGenFunction::emitGeneratorFunction(SILDeclRef function, VarDecl *var) {
+      if (originalProperty->
+          isInnermostPropertyWrapperInitUsesEscapingAutoClosure()) {
+        interfaceType = FunctionType::get({}, interfaceType);
+        varType = FunctionType::get({}, varType);
```
Another place where we could use that new function on `VarDecl`, them `varType` is `mapTypeIntoContext(the-wrappedValue-init-type)`.

DougGregor commented on lib/Sema/TypeCheckPropertyWrapper.cpp:
```diff
@@ -650,16 +678,25 @@ Expr *swift::buildPropertyWrapperInitialValueCall(
+      auto paramsList = init->getParameters();
+      if (paramsList->size() > 0 && paramsList->get(0)->isAutoClosure()) {
+        auto fnTy = FunctionType::get({}, resultTy);
+        initializer = new (ctx) AutoClosureExpr(initializer, fnTy,
```
I'm surprised that you need to create the `AutoClosureExpr` at all here, because usually the type checker would introduce it itself. Do we need this particular code?

roop replied:
Indeed, it appears that we don't need this code. I'll remove it. (I didn't know it would be taken care of by the type checker.)

DougGregor replied:
Great!

DougGregor commented on lib/Sema/TypeCheckStorage.cpp:
```diff
@@ -2392,8 +2392,11 @@ PropertyWrapperBackingPropertyInfoRequest::evaluate(Evaluator &evaluator,
   // original property's type.
+  Type origValueType = var->getType();
+  if (var->isInnermostPropertyWrapperInitUsesEscapingAutoClosure())
+    origValueType = FunctionType::get({}, origValueType);
```
Another place were we could use that new API on `VarDecl` to avoid repeating this logic.

DougGregor reviewed (changes requested):
This is looking really good, and I like the approach you're taking here. My main request is to encapsulate that logic for "adjust to a function type", because we're likely to have a maintenance burden going forward if we have to remember to do this adjustment in many places.

roop commented:
With commit 81c0a0e ("Allow init(wrappedValue:) initializers with defaulted args"), the `PropertyWrapperTypeInfo` struct's wrappedValueInit has changed from a `ConstructorDecl` pointer to an `enum`. With `wrappedValueInit` as a `ConstructorDecl`, I could examine that to see if `wrappedValue:` is an `@autoclosure`, but with the enum, I cannot.

I'm thinking I should add the info of whether there's an `@autoclosure` as a member of `PropertyWrapperTypeInfo` itself.

I'll create a fresh commit set for that incorporating all review comments raised here and force-push to this PR.

roop commented:
Sorry about the delay.

I just force-pushed a fresh set of commits that looks a lot cleaner, which has the following additional changes:

  - Incoporates @DougGregor's suggestion to have a new API in VarDecl. It’s called `getPropertyWrapperInitValueInterfaceType()`.
  - Adds a new var `isWrappedValueInitUsingEscapingAutoClosure` to `PropertyWrapperTypeInfo`, and populates it in the request
  - Since the property wrapper initializer call is now spun off into a separate function in the raw SIL (commit ab5d161c059), there are new places where we should handle the type of the `wrappedValue:` arg:
	- `getPropertyWrapperBackingInitializerInterfaceType` in TypeLowering.cpp
        - `emitGeneratorFunction(SILDeclRef function, Expr *value, bool EmitProfilerIncrement)` in SILGenFunction.cpp

However, this now crashes for the case where the wrapped property is assigned to inside an initializer:

``` Swift
struct UseLazy {
    @Lazy var number: Int // Lazy is a property wrapper using autoclosure
    init() {
        number = 42
    }
}
```

The crash is because SIL verification failed. The “number = 42” translates to an [assign_by_wrapper] instruction in the raw SIL. The `init` argument of that instruction takes an `() -> Int` (which comes from `getPropertyWrapperBackingInitializerInterfaceType`), but the `set` argument of that instruction takes an `Int`, causing a type mismatch:

```
assign_by_wrapper %7 : $Int to %9 : $*Lazy<Int>, init %11 : $@callee_guaranteed (@owned @callee_guaranteed () -> Int) -> @owned Lazy<Int>, set %13 : $@callee_guaranteed (Int) -> ()
```

To compare, the raw SIL without this PR would look like:

```
assign_by_wrapper %7 : $Int to %9 : $*Lazy<Int>, init %11 : $@callee_guaranteed (Int) -> @owned Lazy<Int>, set %13 : $@callee_guaranteed (Int) -> ()
```

To solve this, I think I should do this:

1. Create a separate SIL function that wraps the setter, but takes a closure argument. It calls the closure and passes the result to the setter.
2. Pass this separate function as the `set` argument in the `assign_by_wrap` instruction

Do you think that’s the correct approach?

[assign_by_wrapper]: https://github.com/apple/swift/blob/master/docs/SIL.rst#assign-by-wrapper

roop committed:
[Property wrappers] Fix handling of @autoclosure in init(wrappedValue:)

If the 'wrappedValue:' parameter is an escaping autoclosure, and a
struct property is marked with that property wrapper, the memberwise
initializer of the struct is now synthesized with an escaping
autoclosure for that property.

roop committed:
[Property wrappers] Avoid assign_by_wrapper when autoclosure is involved

When assigning to a wrapped variable in 'init()', if the property wrapper's
'wrappedValue:' parameter is an escaping autoclosure, the initializer and
setter have incompatible types -- the setter takes a value, and initializer
takes a closure returning that value.

An assign_by_wrapper with incompatible types causes SIL verification to fail.

This commit makes the SIL not use assign_by_wrapper in such cases, and use the
setter directly, resulting in a definitive initialization (DI) error instead.

    struct S {
      @Lazy var n: Int // Lazy.init(wrappedValue: @autoclosure ... )
      init() {
        n = 1 // error: 'self' used before all stored properties are initialized
      }
    }

roop committed:
[Property wrappers] Rename hasPropertyWrapper() for clarity

GetterSetterComponent::hasPropertyWrapper() now checks for more
stuff than just whether there's a property wrapper or not.

It's therefore renamed to canRewriteSetAsPropertyWrapperInit().

roop committed:
[Property wrappers] Update existing tests for @autoclosure handling

roop committed:
[Property wrappers] New tests for @autoclosure handling

roop commented:
> To solve this, I will try to implement the following:
> 
> 1. Create a separate SIL function that wraps the setter, but takes a closure argument. It calls the closure and passes the result to the setter.
> 2. Pass this separate function as the `set` argument in the `assign_by_wrap` instruction

In my understanding, implementing that involves creating an alternate setter. I thought that was not worth the complexity.

Instead, commit 58f60a4 causes the raw SIL to skip the `assign_by_wrapper` and call the setter directly, which is caught as a Definitive Initialization error when creating the canonical SIL. Do you think this is acceptable?

@DougGregor: Please review.
Cc: @theblixguy

CodaFi commented on lib/SILGen/SILGenFunction.cpp:
```diff
@@ -813,8 +814,8 @@ void SILGenFunction::emitGeneratorFunction(SILDeclRef function, VarDecl *var) {
-      interfaceType = originalProperty->getValueInterfaceType();
-      varType = originalProperty->getType();
+      interfaceType = originalProperty->getPropertyWrapperInitValueInterfaceType();
+      varType = originalProperty->getDeclContext()->mapTypeIntoContext(interfaceType);
```
`VarDecl::getType` already does this.

CodaFi commented:
@swift-ci please smoke test

roop commented:
I force-pushed to fix merge conflicts.

sindresorhus commented:
@DougGregor Any chance this will make it into Swift 5.2?

futurejones commented:
Can one of the admins verify this patch?

DougGregor committed:
Merge branch 'master' into sr10950_prop_wrap_autoclosure

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
FWIW, I've taken over this pull request via https://github.com/apple/swift/pull/30537 to help get it landed. There were some merge conflicts and I found some cases where it wasn't quite working properly, so those are addressed over there. I recommend you close this PR and we can do follow-ups if there's anything we missed. Thank you so much for this work, @roop, and I'm sorry it took so long.

roop commented:
Awesome. Thanks for taking it over, finishing up and landing it.
